File: src/test/java/com/fwdekker/randomness/ui/ButtonGroupHelperTest.java
package com.fwdekker.randomness.ui;

import kotlin.Unit;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import javax.swing.AbstractButton;
import javax.swing.ButtonGroup;
import javax.swing.JButton;
import java.util.NoSuchElementException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;


/**
 * Unit tests for {@link ButtonGroupHelper}.
 */
final class ButtonGroupHelperTest {
    private ButtonGroup group;


    @BeforeEach
    void beforeEach() {
        group = new ButtonGroup();
    }


    @Test
    void testForEachEmpty() {
        final int[] sum = {0};

        ButtonGroupHelper.INSTANCE.forEach(group, button -> {
            sum[0]++;
            return Unit.INSTANCE;
        });

        assertThat(sum[0]).isEqualTo(0);
    }

    @Test
    void testForEach() {
        final AbstractButton buttonA = new JButton();
        final AbstractButton buttonB = new JButton();
        final AbstractButton buttonC = new JButton();

        group.add(buttonA);
        group.add(buttonB);
        group.add(buttonC);

        final int[] sum = {0};

        ButtonGroupHelper.INSTANCE.forEach(group, button -> {
            sum[0]++;
            return Unit.INSTANCE;
        });

        assertThat(sum[0]).isEqualTo(3);
    }


    @Test
    void testGetValueEmpty() {
        assertThat(ButtonGroupHelper.INSTANCE.getValue(group)).isNull();
    }

    @Test
    void testGetValueNoneSelected() {
        final AbstractButton button = new JButton();

        group.add(button);

        assertThat(ButtonGroupHelper.INSTANCE.getValue(group)).isNull();
    }

    @Test
    void testGetValue() {
        final AbstractButton buttonA = new JButton();
        buttonA.setActionCommand("29zo4");
        final AbstractButton buttonB = new JButton();
        buttonB.setSelected(true);
        buttonB.setActionCommand("Y6ddy");

        group.add(buttonA);
        group.add(buttonB);

        assertThat(ButtonGroupHelper.INSTANCE.getValue(group)).isEqualTo("Y6ddy");
    }


    @Test
    void testSetValueEmpty() {
        assertThatThrownBy(() -> ButtonGroupHelper.INSTANCE.setValue(group, "syWR#"))
            .isInstanceOf(NoSuchElementException.class)
            .hasMessage("Could not find a button with action command `syWR#`.");
    }

    @Test
    void testSetValueNotFound() {
        final AbstractButton buttonA = new JButton();
        buttonA.setActionCommand("*VgyA");
        final AbstractButton buttonB = new JButton();
        buttonB.setActionCommand("s8vOP");

        group.add(buttonA);
        group.add(buttonB);

        assertThatThrownBy(() -> ButtonGroupHelper.INSTANCE.setValue(group, "OD>5&"))
            .isInstanceOf(NoSuchElementException.class)
            .hasMessage("Could not find a button with action command `OD>5&`.");
    }

    @Test
    void testSetValue() {
        final AbstractButton buttonA = new JButton();
        buttonA.setActionCommand("TRUaN");
        final AbstractButton buttonB = new JButton();
        buttonB.setActionCommand("2Y@2_");
        final AbstractButton buttonC = new JButton();
        buttonC.setActionCommand("#Oq%n");

        group.add(buttonA);
        group.add(buttonB);
        group.add(buttonC);

        ButtonGroupHelper.INSTANCE.setValue(group, "2Y@2_");

        assertThat(buttonB.isSelected()).isTrue();
    }

    @Test
    void testSetValueObject() {
        final AbstractButton buttonA = new JButton();
        buttonA.setActionCommand("iqGfVwJDLd");
        final AbstractButton buttonB = new JButton();
        buttonB.setActionCommand("ouzioKGsKi");
        final AbstractButton buttonC = new JButton();
        buttonC.setActionCommand("pKVEAoQzmr");

        group.add(buttonA);
        group.add(buttonB);
        group.add(buttonC);

        ButtonGroupHelper.INSTANCE.setValue(group, new Object() {
            @Override
            public String toString() {
                return "ouzioKGsKi";
            }
        });

        assertThat(buttonB.isSelected()).isTrue();
    }
}

File: src/test/kotlin/com/fwdekker/randomness/ui/ButtonGroupHelperTest.kt
package com.fwdekker.randomness.ui

import org.assertj.core.api.Assertions.assertThat
import org.assertj.swing.edt.FailOnThreadViolationRepaintManager
import org.assertj.swing.edt.GuiActionRunner
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import javax.swing.ButtonGroup
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JRadioButton


/**
 * Unit tests for the extension functions in `ButtonGroupKt`.
 */
object ButtonGroupHelperTest : Spek({
    lateinit var group: ButtonGroup


    beforeGroup {
        FailOnThreadViolationRepaintManager.install()
    }

    beforeEachTest {
        group = ButtonGroup()
    }


    describe("forEach") {
        it("iterates 0 times over an empty group") {
            var sum = 0
            group.forEach { sum++ }

            assertThat(sum).isEqualTo(0)
        }

        it("iterates once for each button in a group") {
            group.add(createJButton())
            group.add(createJButton())
            group.add(createJButton())

            var sum = 0
            group.forEach { sum++ }

            assertThat(sum).isEqualTo(3)
        }
    }

    describe("getValue") {
        it("returns null if the group is empty") {
            assertThat(group.getValue()).isNull()
        }

        it("returns null if no button is selected") {
            group.add(createJButton(actionCommand = "medicine"))

            assertThat(group.getValue()).isNull()
        }

        it("returns an empty string if the selected button has an empty action command") {
            group.add(createJButton(actionCommand = "", isSelected = true))

            assertThat(group.getValue()).isEmpty()
        }

        it("returns the action command of the selected button") {
            group.add(createJButton(actionCommand = "frequent"))
            group.add(createJButton(actionCommand = "umbrella", isSelected = true))

            assertThat(group.getValue()).isEqualTo("umbrella")
        }
    }

    describe("setSalue") {
        it("deselects the currently selected button if no button has the given action command") {
            val buttonA = createJButton(actionCommand = "plenty")
            val buttonB = createJButton(actionCommand = "date")
            val buttonC = createJButton(actionCommand = "bath", isSelected = true)

            group.add(buttonA)
            group.add(buttonB)
            group.add(buttonC)

            GuiActionRunner.execute { group.setValue("formal") }

            assertThat(buttonC.isSelected).isFalse()
        }

        it("selects the button with the given action command") {
            val buttonA = createJButton(actionCommand = "cape")
            val buttonB = createJButton(actionCommand = "another")
            val buttonC = createJButton(actionCommand = "empire")

            group.add(buttonA)
            group.add(buttonB)
            group.add(buttonC)

            GuiActionRunner.execute { group.setValue("another") }

            assertThat(buttonB.isSelected).isTrue()
        }

        it("selects exactly one button if multiple buttons have the given action command") {
            val buttonA = createJButton(actionCommand = "secrecy")
            val buttonB = createJButton(actionCommand = "secrecy")
            val buttonC = createJButton(actionCommand = "study")

            group.add(buttonA)
            group.add(buttonB)
            group.add(buttonC)

            GuiActionRunner.execute { group.setValue("secrecy") }

            assertThat(buttonA.isSelected).isNotEqualTo(buttonB.isSelected)
        }
    }

    describe("buttons") {
        it("returns nothing if the group is empty") {
            assertThat(group.buttons()).isEmpty()
        }

        it("returns the elements of the group") {
            val buttonA = createJButton(actionCommand = "whole")
            val buttonB = createJButton(actionCommand = "melt")
            val buttonC = createJButton(actionCommand = "envy")

            group.add(buttonA)
            group.add(buttonB)
            group.add(buttonC)

            assertThat(group.buttons()).containsExactly(buttonA, buttonB, buttonC)
        }
    }

    describe("setLabel") {
        it("sets the `labelFor` if another button is selected") {
            val button1 = GuiActionRunner.execute<JRadioButton> { JRadioButton() }
            val button2 = GuiActionRunner.execute<JRadioButton> { JRadioButton() }
            group.add(button1)
            group.add(button2)

            val label = GuiActionRunner.execute<JLabel> { JLabel() }
            group.setLabel(label)
            GuiActionRunner.execute { button2.isSelected = true }

            assertThat(label.labelFor).isEqualTo(button2)
        }
    }
})


private fun createJButton(actionCommand: String? = null, isSelected: Boolean = false) =
    GuiActionRunner.execute<JButton> {
        JButton().also {
            it.actionCommand = actionCommand
            it.isSelected = isSelected
        }
    }
